ttnn/cpp/ttnn/operations/experimental/ccl/ccl_op_fusion.hpp

```diff
@@ -24,91 +31,84 @@ struct AllGatherFusedOpSignaler {
     bool initialized_fused_op = false;
     bool initialized_all_gather = false;
 
-
-    AllGatherFusedOpSignaler(
-        const std::vector<CoreCoord>& fused_op_receiver_cores,
-        const std::vector<uint32_t>& fused_op_receiver_signal_semaphores)
-        : fused_op_receiver_cores(fused_op_receiver_cores),
-            fused_op_receiver_signal_semaphores(fused_op_receiver_signal_semaphores) {
-
-        }
+    AllGatherFusedOpSignaler() {}
 
     void init_fused_op(
-        Device const* device
-    ) {
-        // Get the noc coords for the fused op receiver cores
-        this->fused_op_receiver_cores_noc.clear();
-        for (const auto& core : this->fused_op_receiver_cores) {
-            this->fused_op_receiver_cores_noc.push_back(device->worker_core_from_logical_core(core));
-        }
-        initialized_fused_op = true;
-    }
+        const std::vector<CoreCoord>& fused_op_receiver_cores_noc,
+        const std::vector<uint32_t>& fused_op_receiver_signal_semaphores
+    );
 
     void init_all_gather(
         Program& program,
         Device const* device,
 
         CoreRangeSet const& all_gather_workers,
         std::vector<CoreCoord>& all_gather_worker_cores
-    ) {
-        // Create the sync semaphore for the all gather workers
-        this->all_gather_worker_sync_semaphore = CreateSemaphore(program, all_gather_workers, 0);
-
-        // Get the noc coords for the all gather workers
-        this->all_gather_worker_cores_noc.clear();
-        for (const auto& core : all_gather_worker_cores) {
-            this->all_gather_worker_cores_noc.push_back(device->worker_core_from_logical_core(core));
-        }
-        initialized_all_gather = true;
-    }
+    );
 
     void emit_all_gather_fused_op_ct_args(
         std::vector<uint32_t>& ct_args,
 
         uint32_t num_workers_to_sync,
         uint32_t curr_worker_index
-    ) {
-        TT_ASSERT(initialized_fused_op && initialized_all_gather, "AllGatherFusedOpSignaler not initialized fully.");
-
-        ct_args.push_back(static_cast<uint32_t>(num_workers_to_sync));
-        ct_args.push_back(static_cast<uint32_t>(curr_worker_index));
-        ct_args.push_back(static_cast<uint32_t>(this->all_gather_worker_sync_semaphore));
-
-    }
+    );
 
 
     void emit_all_gather_fused_op_rt_args(
         std::vector<uint32_t>& rt_args,
 
-        bool all_gather_direction
-    ) {
-        TT_ASSERT(initialized_fused_op && initialized_all_gather, "AllGatherFusedOpSignaler not initialized fully.");
+        uint32_t all_gather_direction,
+        std::optional<CoreSemPair> start_signal_core_sem_pair = {}
+    );
+
+    static uint32_t get_num_ct_args() { return 3; }
+
+};
+
+// Used to propagate semaphore information from matmul to all_gather in all_gather_matmul op
+struct MatmulFusedOpSignaler {
```

Comment:
should matmul specific signaler be out of here in its own file?

Reply:
Currently this file is supposed to hold the different types of signalers required for op fusion (currently there's only an implementation for all_gather+matmul). I can change the file name to `op_fusion.hpp` to make it more semantic. thoughts @SeanNijjar?

Reply:
yea I think this fusion component will be more generally applicable and we can see that actually, besides the weights width, there is nothing matmul specific about it. I think ultimately mostly a "fused op consumer" signal receiver rather than specifically for matmul.

That being said, this is all in experimental and we're still exploring and figuring things out. I do foresee in the future that, as opposed to how the full tensor is the interface to the op, when we do fusion, we'll have something like tensor slices as the interface between the fused op producer and consumer, and this component mostly has just that info.

If anything I think we should keep this here but just rename it. Maybe make a matmul specialization that has the width attribute and it can specialize it locally...